File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/annotations/composer/StatelessInitializer.java
package io.github.oleksivio.telegram.bot.core.controller.annotations.composer;

import io.github.oleksivio.telegram.bot.core.controller.handler.Handler;
import io.github.oleksivio.telegram.bot.core.controller.processor.StatelessProcessor;
import org.springframework.stereotype.Controller;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Optional;

@Controller
public class StatelessInitializer {
    private final HandlerCreator handlerCreator;
    private final StatelessProcessor statelessProcessor;

    public StatelessInitializer(HandlerCreator handlerCreator, StatelessProcessor statelessProcessor) {
        this.handlerCreator = handlerCreator;
        this.statelessProcessor = statelessProcessor;
    }

    public void init(Class<?> objClz, Object obj) {
        Collection<Handler> simpleHandlerList = new ArrayList<>();

        for (Method method : objClz.getDeclaredMethods()) {
            Optional<Handler> optionalHandler = handlerCreator.create(obj, method);
            if (!optionalHandler.isPresent()) {
                continue;
            }
            simpleHandlerList.add(optionalHandler.get());
        }

        for (Handler handler : simpleHandlerList) {
            statelessProcessor.add(handler);
        }
    }
}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/annotations/composer/StatelessInitializer.kt
package io.github.oleksivio.telegram.bot.core.controller.annotations.composer

import io.github.oleksivio.telegram.bot.core.controller.processor.StatelessProcessor
import io.github.oleksivio.tl.kbot.server.api.model.ITelegram
import org.springframework.stereotype.Controller
import kotlin.reflect.KClass
import kotlin.reflect.full.functions

@Controller
class StatelessInitializer(private val handlerCreator: HandlerCreator, private val statelessProcessor: StatelessProcessor) {

    fun init(objClz: KClass<*>, obj: Any) = objClz.functions
            .mapNotNull { handlerCreator.create<ITelegram>(obj, it) }
            .forEach { statelessProcessor.add(it) }

}
